SeaweedFS blob storage

kiendang opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

shubham3121 committed:
set blob storage base directory to ~/.syft/data/<node.id>

shubham3121 committed:
create base directory if not exists during OnDisk client initialization

shubham3121 committed:
add seaweedfs to docker compose
use leveldb instead of redis

shubham3121 committed:
- add boto3 to setup.cfg
- add connect method to BlobStorageConnection
- implement connect method for OnDisk storage connection
- define SeaweedFS client config
- implement SeaweedFSClient using boto3
- implement write method in SeaweedFSBlobDeposit
- implement read, allocate and write methods in SeaweedFSConnection
- add service method to mark multipart upload complete
- add method to mark multipart upload complete in SeaweedFSConnection.

Co-authored-by: Kien Dang <kiendang@users.noreply.github.com>
Co-authored-by: Khoa Nguyen <dkn.work@protonmail.com>

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
Minor fixes/cleanup

kiendang committed:
Also catch ConnectionError in addition to HTTPError

kiendang committed:
Specify request timeout

kiendang committed:
Use Path.mkdir

kiendang committed:
Create parent directories of .syft/data if needed

kiendang committed:
Fix util.download_file

not related to seaweed but this is right below the code
I was working on

kiendang committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
-  create seaweed client config in grid/node.py
- create and pass blob storage config to Domain and Gateway in grid/node.py
- Define BlobDepositType typing
- fix typo in name for mark_write_complete service
- temporary convert s3 url host to localhost/blob
- fix size calculation of obj in CreateBlobStorageEntry
- create a separate secure location class for SeaweedFS Client

Co-authored-by: Khoa Nguyen <dkn.work@protonmail.com>
Co-authored-by: Kien Dang <kiendang@users.noreply.github.com>
Co-authored-by: Peter Chung <96817160+PeterChung241@users.noreply.github.com>

kiendang committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
- add route for Blob Store in Routes
- add method to SyftClient that maps a given path to blob route
- save pre-signed urls as GridUrl objects
- convert pre-signed urls to /blob route on client side during read and write

kiendang committed:
Fix GridURL typing

and a minor bug copy -> __copy__

shubham3121 committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
- pass s3 username and password via hagrid as cli agruments
- pass s3 username and password to s3config.json via jinja template
- fix usage of templates when running hagrid in EDITABLE_MODE

shubham3121 committed:
revert to use static s3config.json

shubham3121 committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
Initial work on action object integration

Co-authored-by: Peter Chung <PeterChung241@users.noreply.github.com>
Co-authored-by: Khoa Nguyen <dkn.work@protonmail.com>
Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

shubham3121 committed:
update seaweed container command to configure s3 credentials dynamically

khoaguin committed:
fix some linting issues

shubham3121 committed:
- revert delete_data logic

kiendang committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
- rename __set_syft_action_data_cache to _set_syft_action_data
- serialize data to bytes before saving it to seaweed in _set_syft_action_data
- change typing for type_ field in BlobStorageEntry and CreateBlobStorageEntry

kiendang committed:
Updates

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

shubham3121 committed:
hide std output of the s3 config credentials

shubham3121 committed:
create a start script for seaweedfs and move the entrypoint to it
add seaweedfs start script to manifest

PeterChung241 committed:
Added delete method for service and on_disk

PeterChung241 committed:
added delete method to seaweedfs

shubham3121 committed:
remove print statements
add syft_action_data and _set_syft_action_data to pass through attrs
remove unused validator

kiendang committed:
Merge branch 'dev' into seaweed-integration

PeterChung241 committed:
Added delete method for SeaweedFSConnection

shubham3121 committed:
- replace checking type from syft_action_data to self.syft_action_data_type
- replace checking bool type on syft_action_data to self.syft_has_bool_attr

kiendang commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -881,8 +942,8 @@ def __post_init__(self) -> None:
 
-        if isinstance(self.syft_action_data, ActionObject):
-            raise Exception("Nested ActionObjects", self.syft_action_data)
+        if isinstance(self.syft_action_data_type, ActionObject):
```
should this be `issubclass` instead of `isinstance` since `syft_action_data_type` is a type?

shubham3121 committed:
- add AuthNodeContextRegistry class
- use auth node context in syft action data is API is done
- update blob api to use node context if api is not None
- save result object on execute in action service

shubham3121 committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
integrate blob upload save method with Twin Object

PeterChung241 committed:
Added delete to BlobStorageConnection

PeterChung241 committed:
updated delete methods to return more specific errors

shubham3121 committed:
create a method to get from context or api
update syft_action_data and _set_syft_action_data to use from_context_or_api

shubham3121 committed:
move AuthNodeContextRegistry.set_node_context to api call method
fix Action Object creation in convert_to_pointers
add __check_action_data to BASE_PASSTHROUGH_ATTRS

shubham3121 committed:
fix on disk write

shubham3121 committed:
Merge pull request #8002 from PeterChung241/seaweed-integration

Added delete methods for Blob Storage

shubham3121 committed:
skip serde for syft_action_data_cache in Pandas Action Object

kiendang committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
- add a created time DateTime field to ActionObject
- rename .save method to _save_to_blob_store
- conditionally fetch data from blob store in case of syft_action_data
- set syft_action_data to ActionDataEmpty post saving data to blob store

shubham3121 committed:
skip syft_action_data_cache check if value is ActionEmptyData

shubham3121 committed:
- Added a method to return ActionEmptyData object at ActionObject level
- propogate repr, str, type and bool information for Action Objects

Co-authored-by: Kien Dang <kiendang@users.noreply.github.com>
Co-authored-by: Peter Chung <PeterChung241@users.noreply.github.com>

shubham3121 committed:
- rename .save in twin object to _save_to_blob_store
- replace action.save method with action.save in requests.py
-  pass syft client verify key and node location to from_obj constructor

Co-authored-by: Kien Dang <mail@kien.ai>

shubham3121 committed:
handle action object repr in _repr_debug

shubham3121 committed:
add a notebook to demo dataset

shubham3121 committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
add missing syft node location and client verify key during Action Object creation

shubham3121 committed:
- change blob service write, read APIs role access to GUEST_ROLE
- handle obj not exists in blob storage service

kiendang committed:
Create blob storage entry from file

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

shubham3121 committed:
- implement a classmethod to return node context registry via userverifykey
- add a method to BaseSyftObject to set node uid and syft client verify key
- fix passing of client verify key and node uid via SyftAPI object
- set syft verify key and node location on Action objs at service level
- add blob storage permission change to ActionStoreChange
- add method for adding and removing permission at BaseStash
- add abstract methods for adding and removing permission to StorePartition

kiendang committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
- replace use of action_store.set to action_service.set
- pass has permission flag via context to action set method
- add extra_kwargs dict to Auth Node Context
- add classmethod decorator to auth_context_for_user
- comment saving mock to blob store

shubham3121 committed:
pass has_result_read_permission from context to action store

shubham3121 committed:
- check if blob storage permission before removing during undo request
- implement has_permission method to document store and base stash

shubham3121 committed:
fix import in blob storage notebook

shubham3121 committed:
fix issues of blob storage with enclave service
fix issues with execute code
handle error at blob storage
add missing permissions for blob storage

shubham3121 committed:
update syft action data to handle file type uploads to blob store

shubham3121 committed:
pass blob storage config to worker settings
reset blob storage if node is reset

shubham3121 committed:
fix key generation in AuthNodeContextRegistry

shubham3121 committed:
add serde for blob storage config
fix type for blob storage config in worker settings

shubham3121 committed:
unlink blob directory only if it exists

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
CI

shubham3121 committed:
- fetch api for root and guest client in node.py
- fix repr for action objects
- fix filter for args and kwargs during Plan execute

teo-milea committed:
Merge branch 'dev' into seaweed-integration

koenvanderveen committed:
fix part of the tests

koenvanderveen committed:
fix more tests

koenvanderveen committed:
more tests

shubham3121 committed:
fix type checking in is_action_data_empty function

shubham3121 committed:
clean up prints

shubham3121 committed:
Merge branch 'seaweed-integration' into seaweed-bugfix

shubham3121 committed:
Merge pull request #8050 from OpenMined/seaweed-bugfix

WIP: Seaweed bugfix

kiendang committed:
Fix notebook

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
Fix user_test

Update services.user.update to use auto_splat syntax

kiendang committed:
Revert services.user.update autosplat

kiendang commented on packages/syft/src/syft/service/user/user_service.py:
```diff
@@ -210,7 +210,12 @@ def get_current_user(
+        path="user.update",
+        name="update",
+        roles=GUEST_ROLE_LEVEL,
+        autosplat=["user_update"],
```
@koenvanderveen I reverted this for now. This breaks `user_test.py` even after updating the syntax. `UserUpdate` has a couple of validators which don't get applied if using `autosplat`.

https://github.com/OpenMined/PySyft/blob/687aefd64076f312e26ebd855221baf200d418f2/packages/syft/src/syft/service/user/user.py#L114-L122

koenvanderveen replied:
👍

teo-milea committed:
Merge branch 'seaweed-integration' of github.com:OpenMined/PySyft into seaweed-integration

kiendang committed:
EmailStr is str

kiendang committed:
Remove unused definition

kiendang committed:
More correct typing

teo-milea committed:
Merge branch 'seaweed-integration' of github.com:OpenMined/PySyft into seaweed-integration

kiendang committed:
Attempt to fix RecursionError

kiendang committed:
Merge branch 'dev' into seaweed-integration

teo-milea committed:
Merge branch 'seaweed-integration' of github.com:OpenMined/PySyft into seaweed-integration

teo-milea committed:
fix numpy tests

kiendang committed:
Revert

koenvanderveen committed:
fix some tests

koenvanderveen committed:
Merge branch 'seaweed-integration' of github.com:OpenMined/PySyft into seaweed-integration

shubham3121 committed:
revert user changes

koenvanderveen committed:
fix action object tests

shubham3121 committed:
set dtype of action object if set to None

koenvanderveen committed:
merge

koenvanderveen committed:
fix conflict

koenvanderveen committed:
worker tests

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
nit: issubclass instead of exact type comparison

kiendang committed:
Delete debug notebook

kiendang committed:
Merge branch 'dev' into seaweed-integration

madhavajay committed:
Pinned seaweedfs version

- Enabled dev mode ports

kiendang committed:
Resource safety

kiendang committed:
Add example uploading retrieving file with blob storage

kiendang committed:
Delete BlobStorageEntry after notebook runs

kiendang committed:
Also delete permission when deleting BlobStorageEntry

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
Merge branch 'dev' into seaweed-integration

kiendang committed:
Merge branch 'dev' into seaweed-integration

koenvanderveen commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -436,29 +468,97 @@ class ActionObject(SyftObject):
+                self.syft_action_data_cache = blob_retrieval_object.read()
+                self.syft_action_data_type = type(self.syft_action_data)
+
+    def _set_syft_action_data(self, data: Any) -> None:
```
I think if we could rename this, it basically just uploads to the blob store. The fact that it then sets the data is maybe less important, especially because its only used in _save_to_blob_store, which just takes the existing data from the object and then passes it into this method.

kiendang replied:
Agree. We could change the name of this. Initially this was created to be the setter for `syft_action_data` as you can see the line right after this function `syft_action_data = syft_action_data.setter(_set_syft_action_data)`. As we don't set `syft_action_data` directly now but through `syft_action_data_cache` we should delete this and rename the function.

There is `def delete_data(self)` which sets `self.syft_action_data = empty`. I don't see it being used anywhere. Should we remove it or modify it (so that it sets `syft_action_data_cache` and removes from blob storage as well)?

shubham3121 replied:
I think we should remove this method as it creates confusion and doesn't perform the functionality as it suggests.

koenvanderveen commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -474,6 +574,40 @@ def make_id(cls, v: Optional[UID]) -> UID:
+            values["syft_has_bool_attr"] = hasattr(v, "__bool__")
+        return values
+
+    def _save_to_blob_store(self) -> Optional[SyftError]:
```
I think this always coupled with client.api.services.action.set, maybe we should have method that does both? I am still thinking about the coupling of these two, because if you only call client.api.services.action.set OR \_save_to_blob_store, and not the other, you end up in pretty weird state. One solution could be to have an endpoint that does both, and make the two individual ones "private" (prefix with '_')

koenvanderveen commented on packages/syft/src/syft/store/blob_storage/seaweedfs.py:
```diff
@@ -1,43 +1,229 @@
+                message=f"Failed to allocate space for {obj} with error: {e}"
+            )
 
+    def write(self, obj: BlobStorageEntry) -> BlobDeposit:
```
maybe rename this method? I think "write" gives a different expectation, especially since we write the result again, resulting in this pattern

```
obj = ...
res = client.write(obj)
res.write()
```

kiendang replied:
Ya these method names are not precise and definitely bikeshed-able but not a huge deal to me tbh.
The `read` and `write` methods under Connection create an object to be sent to the client which they can use to retrieve/deposit a file or `SyftObject` from/to blob storage. Could be `(make/prepare/to)_read/write` but these are not precise either.

koenvanderveen commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -436,29 +468,97 @@ class ActionObject(SyftObject):
     _syft_post_hooks__: Dict[str, List] = {}
     syft_twin_type: TwinMode = TwinMode.NONE
     syft_passthrough_attrs = BASE_PASSTHROUGH_ATTRS
+    syft_action_data_type: Optional[Type]
```
what is the difference between syft_internal_type and syft_action_data_type

shubham3121 replied:
As of now `syft_action_data_type` always stores the type of the data being stored and is calculated on the fly. While what I can see `syft_internal_type` is set explicitly during different Action Object subclasses like `NumpyBoolObject`, `NumpyScalarObject`, etc.

koenvanderveen commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -579,6 +713,19 @@ def _syft_try_to_save_to_store(self, obj) -> None:
         from ...client.api import APIRegistry
 
+        if obj.syft_node_location is None:
+            obj.syft_node_location = obj.syft_node_uid
```
what is the difference between syft_node_location and syft_node_uid?

kiendang replied:
Doesn't look like there's a difference to me either. Will discuss with Shubham on this and other comments. Thanks for the review!

shubham3121 replied:
There is not much difference, `syft_node_location` is set everytime an object is passed through the service_method decorator. While `syft_node_uid` is used to store the id of the node when its saved. Since both of them were developed in parallel, there is a slight redundancy in their use. I agree we should combine them, but would prefer to do that in a seperate PR.

shubham3121 committed:
add Action Object creation from file

kiendang committed:
Rename from_file -> from_path to be more in line with other methods

kiendang committed:
Delete file downloaded and file stored in blob storage after running notebook

kiendang committed:
Merge branch 'dev' into seaweed-integration

shubham3121 committed:
Merge branch 'dev' into seaweed-integration

madhavajay committed:
dev mode True has to trigger local template

- if any environment variables of docker compose changes occur
  the template must be local to the source

- changed docker volume from _ to - naming
- fixed left over action.save command in notebook
- added blob storage path print for debugging in database

madhavajay committed:
Merge pull request #8060 from madhavajay/madhava/seaweed_fixes

Fixes to SeaweedFS PR

teo-milea commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -474,6 +575,40 @@ def make_id(cls, v: Optional[UID]) -> UID:
+            return result
+        if not TraceResult.is_tracing:
+            self.syft_action_data_cache = self.as_empty_data()
+
```
I don't understand the purpose of emptying the cache here. What is TraceResult?

kiendang replied:
@koenvanderveen

shubham3121 replied:
TraceResult is used in Plans (check planify method). TraceResult is a global state and when it is enabled, then we're executing a Plan over local variables. At that point we don't want to clear the cache. 

Reason for Setting cache to empty:
This is because, now saving Action Objects include two steps:
- save actual data (in syft_action_data_cache) to blob store
- post that, save the Action Object itself to the Action Store. Now since we don't want to save the actual data present in `syft_action_data_cache` to the ActionStore (as that data is already stored to blob store) we clear it or set it as empty post saving to blob store.

teo-milea commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -474,6 +575,40 @@ def make_id(cls, v: Optional[UID]) -> UID:
+            return data
+        if isinstance(data, ActionDataEmpty):
+            print(f"cannot store empty object {self.id}")
+            return data
```
We should return an error here as well, so we don't have to check everytime if the result is an error or an ActionDataEmpty

shubham3121 committed:
fix referece for action object in accept_by_depositing_result

kiendang commented:
@shubham3121 I did the same `accept_by_depositing_result` fix but got another error in notebook 05 after that

```python
res_ptr = domain_client.code.func(x=x_pointer)
res = res_ptr.get()
res

AttributeError                            Traceback (most recent call last)
Cell In[21], line 2
      1 res_ptr = domain_client.code.func(x=x_pointer)
----> 2 res = res_ptr.get()
      3 res

AttributeError: 'SyftError' object has no attribute 'get'
```

Do you encounter the same thing locally?

teo-milea commented:
I got the same issue as @kiendang locally with the same fix, how did the tests pass?

shubham3121 commented:
yes, getting the same error, that's weird the tests passed. The problem seems to be in `apply_output` in `RepeatedCallPolicy`: outputs[output_arg] is throwing the error.

kiendang commented:
Hmm I thought `outputs[output_arg]` raised the previous error which got fixed with `accept_by_depositing_result`?
The error from `domain_client.code.func(x=x_pointer)` looks like a separate issue.

```python
res_ptr = domain_client.code.func(x=x_pointer)
res_ptr

SyftError: Failed to run. 'NoneType' object has no attribute 'services'
```

teo-milea committed:
used syft_action_data in output policy func

teo-milea committed:
removed debug prints

teo-milea committed:
revert change

teo-milea committed:
changed example policy

teo-milea commented:
I have changed the example policy from notebook 5 as this feature is not really used and we could offer support to anyone trying it out. Given the changes from the last few months, we need to refactor and rethink the UX for custom policies, but it does not feel like a priority right now.

teo-milea commented on packages/syft/src/syft/service/action/action_service.py:
```diff
@@ -71,10 +81,22 @@ def set(
+        else:
+            action_object.private_obj.syft_created_at = DateTime.now()
+            action_object.mock_obj.syft_created_at = DateTime.now()
+
```
Nitpick: Could we be specific here that we check if the object is a TwinObject, and let the default case to be the ActionObject?

teo-milea commented on packages/syft/src/syft/service/request/request.py:
```diff
@@ -97,29 +99,63 @@ def _run(
+            )
+
+            if action_obj.is_err():
+                return Err(SyftError(message=f"{action_obj.err()}"))
```
Nitpick: do we need both Err and SyftErr?

teo-milea commented on packages/syft/src/syft/service/request/request.py:
```diff
@@ -792,7 +841,7 @@ def _run(
             obj = self.linked_obj.resolve_with_context(context)
             if obj.is_err():
-                return SyftError(message=obj.err())
+                return Err(SyftError(message=obj.err()))
```
NP: same problem with Err and SyftError

teo-milea commented on packages/syft/src/syft/service/request/request.py:
```diff
@@ -603,7 +652,7 @@ def _run(
             obj = self.linked_obj.resolve_with_context(context)
             if obj.is_err():
-                return SyftError(message=obj.err())
+                return Err(SyftError(message=obj.err()))
```
NP: same problem with Err and SyftError

shubham3121 replied:
@teo-milea there is a bit inconsistent use of Err and SyftError at request and Action store. Resorting to still use Result library over SyftError to prevent breaking existing functionality. We need to refactor this a bit and make their use consistent all across the codebase.

kiendang commented:
Notebook 06 also fails for me locally.

teo-milea commented:
Works for me, can you share the error?

kiendang commented:
<img width="660" alt="Screenshot 2023-09-01 at 5 43 04 PM" src="https://github.com/OpenMined/PySyft/assets/6521018/393000bf-b5ad-42a4-bd40-4cff60f1b9cb">

shubham3121 commented:
@kiendang notebook 6 is working fine. Can you take the latest pull ?

kiendang commented:
> @kiendang notebook 6 is working fine. Can you take the latest pull ?

Yup it works now after clearing the db from previous runs.

kiendang committed:
Remove unused methods

kiendang committed:
Rename methods

kiendang committed:
Delete debug notebook

kiendang committed:
Return SyftError when attempting to save ActionDataEmpty

madhavajay committed:
Merge branch 'dev' into seaweed-integration

madhavajay committed:
Updated opendp code in blob storage notebook

khoaguin committed:
Merge with `dev` and resolve conflicts in `mongo_document_store.py`n

madhavajay commented:
🚂 Choo Choo

kiendang committed:
Merge branch 'dev' into seaweed-integration
